docs/design_doc.md

```diff
@@ -0,0 +1,118 @@
+
+    * Redis can be configured to be [highly available](https://redis.io/topics/sentinel) with automatic failover. 
+
+    * A Redis cluster provides [weak consistency](https://redis.io/topics/cluster-tutorial#redis-cluster-consistency-guarantees) which is acceptable since any stale cache reads will result in an additional request to the Github API (as well as unnecessarily replacing the cache entry), but the client will not receive stale data.
+
+    * Other useful features of Redis include sharding the cache if it gets too big, as well as the ability to select from a number of cache eviction policies.
+
+    Why AWS Elasticache?
+
+    * Managing a persisent, highly-available and sharded Redis cluster can be greatly simplified using Elasticache, thereby reducing the maintence burden of the github mirror.
+
+    * Since the github mirror OpenShift cluster is running on AWS nodes, the AWS Elasticache Redis cluster can be co-located in the same region to reduce network time.
+
+* Measuring latency
+
+    * A Prometheus metrics endpoint currently provides the latency of requests to the mirror, and is used to display the 90, 95 and 99 percentile tail latencies in Grafana. A separate metric can be added to show the additional latency created by introducing the Redis cache server.
+
+### Architecture diagram
+
+![](./images/arch.png)
+
+[source](./images/arch.svg)
+
+## Scope
+
+* Cache persistence is limited to occasionally backing up the cache contents to prevent the github mirror from starting with an empty cache after a restart.
+
+* Any additional work on metrics for Prometheus or Grafana that is necessary to analyze and evaluate the new design is within the scope of this project.
+
+## Acceptance Criteria
+
+* A successful solution will provide a highly available, low latency mirror server, that can be easily scaled from 3 to 100 replicas without causing the client to be rate limited by the Github API.
+
+* Restarting the github mirror will not cause a large spike in cache misses, if a backup of the cache contents is available. It is expected that a few writes to the cache will be lost in the process, but that is an acceptable outcome and can be adjusted by changing the snapshotting frequency.
+
+* Additional latency due to redesign will not exceed 3 [http request times](http://services.google.com/fh/files/blogs/google_delayexp.pdf).
+
+* Users can set up the mirror to use either the per-pod local in-memory cache or the Redis server simply by modifying a configuration file. 
+
+* If the Redis cache is unavailable, the github mirror can default to using the local in-memory cache.
+
+## Extensions
+
+* [APPSRE-1795](https://issues.redhat.com/browse/APPSRE-1795): Investigate the possibility to use github-mirror as oauth2 auth-provider
```

Comment:
This link is not publicaly accessible. Also, it's an unrelated feature, so I'd just not have it here. An additional proposal can be created for it.

Reply:
ok